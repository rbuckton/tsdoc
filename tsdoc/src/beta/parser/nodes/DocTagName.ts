import { SyntaxKind } from "./SyntaxKind";
import { Syntax, ISyntaxParameters } from "./Syntax";

export interface IDocTagNameParameters extends ISyntaxParameters {
    text?: string;
}

export class DocTagName extends Syntax {
    private _text: string | undefined;

    public constructor(parameters?: IDocTagNameParameters) {
        super(parameters);
        this._text = parameters && parameters.text;
    }

    /**
     * @override
     */
    public get kind(): SyntaxKind.DocTagName {
        return SyntaxKind.DocTagName;
    }

    public get text(): string {
        return this._text || "@unknown";
    }

    public set text(value: string) {
        if (this.text !== value) {
            this.beforeChange();
            this._text = value;
            this.afterChange();
        }
    }
}
