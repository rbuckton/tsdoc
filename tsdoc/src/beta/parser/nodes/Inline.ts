import { INodeParameters } from "./Node";
import { Block } from "./Block";
import { Content, IContentParameters } from "./Content";

export interface IInlineParameters extends IContentParameters {
}

export abstract class Inline extends Content {
    // @ts-ignore
    private _inlineBrand: never;

    constructor(parameters?: INodeParameters) {
        super(parameters);
    }

    /**
     * Gets the parent of this node, if that parent is a `Block`.
     */
    public get parentBlock(): Block | undefined {
        return this.parent && this.parent.isBlock() ? this.parent : undefined;
    }

    /**
     * Gets the parent of this node, if that parent is an `Inline`.
     */
    public get parentInline(): Inline | undefined {
        return this.parent && this.parent.isInline() ? this.parent : undefined;
    }

    /**
     * Gets the previous sibling of this node, if that sibling is an `Inline`.
     */
    public get previousSiblingInline(): Inline | undefined {
        return this.previousSibling && this.previousSibling.isInline() ? this.previousSibling : undefined;
    }

    /**
     * Gets the next sibling of this node, if that sibling is an `Inline`.
     */
    public get nextSiblingInline(): Inline | undefined {
        return this.nextSibling && this.nextSibling.isInline() ? this.nextSibling : undefined;
    }

    /**
     * Gets the first child of this node, if that child is an `Inline`.
     */
    public get firstChildInline(): Inline | undefined {
        return this.firstChild && this.firstChild.isInline() ? this.firstChild : undefined;
    }

    /**
     * Gets the last child of this node, if that child is an `Inline`.
     */
    public get lastChildInline(): Inline | undefined {
        return this.lastChild && this.lastChild.isInline() ? this.lastChild : undefined;
    }

    /**
     * @override
     */
    public isInline(): true {
        return true;
    }
}