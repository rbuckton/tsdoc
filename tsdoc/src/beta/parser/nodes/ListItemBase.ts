import { Block, IBlockParameters } from "./Block";
import { ListBase, IListMarker } from "./ListBase";

export interface IListItemBaseParameters extends IBlockParameters {
}

export abstract class ListItemBase extends Block {
    public constructor(parameters?: IListItemBaseParameters) {
        super(parameters);
    }

    /**
     * Gets the parent of this node, if that parent is a `ListBase`.
     */
    public get parentList(): ListBase | undefined {
        return this.parent && this.parent.isList() ? this.parent : undefined;
    }

    /**
     * Gets the previous sibling of this node, if that sibling is a `ListItemBase`.
     */
    public get previousSiblingListItem(): ListItemBase | undefined {
        return this.previousSibling && this.previousSibling.isListItem() ? this.previousSibling : undefined;
    }

    /**
     * Gets the next sibling of this node, if that sibling is a `ListItemBase`.
     */
    public get nextSiblingListItem(): ListItemBase | undefined {
        return this.nextSibling && this.nextSibling.isListItem() ? this.nextSibling : undefined;
    }

    /**
     * Gets the list marker for this list item.
     */
    public abstract get listMarker(): IListMarker;

    /** @override */
    public isListItem(): true {
        return true;
    }

    /** @override */
    public isBlockContainer(): true {
        return true;
    }
}