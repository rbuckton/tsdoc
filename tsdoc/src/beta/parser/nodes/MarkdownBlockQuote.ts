import { SyntaxKind } from "./SyntaxKind";
import { Block, IBlockParameters } from "./Block";
import { TSDocPrinter } from "../TSDocPrinter";

export interface IMarkdownBlockQuoteParameters extends IBlockParameters {
}

export class MarkdownBlockQuote extends Block {
    public constructor(parameters?: IMarkdownBlockQuoteParameters) {
        super(parameters);
    }

    /** @override */
    public get kind(): SyntaxKind.MarkdownBlockQuote {
        return SyntaxKind.MarkdownBlockQuote;
    }

    /** @override */
    public isBlockContainer(): true {
        return true;
    }

    /** @override */
    protected print(printer: TSDocPrinter): void {
        printer.pushBlock({
            linePrefix: '>'
        });
        this.printChildren(printer);
        printer.popBlock();
    }
}
