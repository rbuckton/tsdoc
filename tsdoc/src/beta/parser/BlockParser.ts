import { Scanner, IScannerState } from "./Scanner";
import { Token } from "./Token";
import { SyntaxKind } from "../nodes/SyntaxKind";
import { StartResult, ContinueResult, IBlockSyntaxParser } from "./blockParsers/IBlockSyntaxParser";
import { DocumentParser } from "./blockParsers/DocumentParser";
import { DocBlockTagParser } from "./blockParsers/DocBlockTagParser";
import { MarkdownBlockQuoteParser } from "./blockParsers/MarkdownBlockQuoteParser";
import { MarkdownHeadingParser } from "./blockParsers/MarkdownHeadingParser";
import { MarkdownCodeBlockParser } from "./blockParsers/MarkdownCodeBlockParser";
import { MarkdownHtmlBlockParser } from "./blockParsers/MarkdownHtmlBlockParser";
import { MarkdownThematicBreakParser } from "./blockParsers/MarkdownThematicBreakParser";
import { MarkdownListParser } from "./blockParsers/MarkdownListParser";
import { MarkdownListItemParser } from "./blockParsers/MarkdownListItemParser";
import { MarkdownParagraphParser } from "./blockParsers/MarkdownParagraphParser";
import { InlineParser } from "./InlineParser";
import { Document } from "../nodes/Document";
import { Node } from "../nodes/Node";
import { MarkdownCodeBlock } from "../nodes/MarkdownCodeBlock";
import { MarkdownParagraph } from "../nodes/MarkdownParagraph";
import { Block } from "../nodes/Block";
import { ContentWriter } from "./ContentWriter";
import { IMapping } from "./Preprocessor";
import { ParserBase } from "./ParserBase";

export class BlockParser extends ParserBase {
    private _line: number;
    private _linePos: number;
    private _indent: number;
    private _blank: boolean;
    private _indentStart: IScannerState | undefined;
    private _lastNextNonIndentPos: number;
    private _previousLineEnd: number;
    private _root!: Document;
    private _lastMatch: Block | undefined;
    private _hasUnmatchedBlocks: boolean;
    private _tip: Block | undefined;
    private _lastTip: Block | undefined;
    private _blockSyntaxParserMap: Map<SyntaxKind, IBlockSyntaxParser<Block>>;
    private _blockSyntaxParsers: ReadonlyArray<IBlockSyntaxParser<Block>>;

    public constructor(text: string, sourceMappings?: IMapping[]) {
        super(text, sourceMappings);
        this._line = 0;
        this._linePos = 0;
        this._indent = 0;
        this._blank = false;
        this._lastNextNonIndentPos= -1;
        this._previousLineEnd= -1;
        this._hasUnmatchedBlocks = false;
        this._blockSyntaxParsers = BlockParser.getDefaultBlockSyntaxParsers();
            this._blockSyntaxParserMap = new Map();
        for (const blockSyntaxParser of this._blockSyntaxParsers) {
            this._blockSyntaxParserMap.set(blockSyntaxParser.kind, blockSyntaxParser);
        }
    }

    /**
     * Gets the state of the scanner before the current indent.
     */
    public get indentStartState(): IScannerState | undefined {
        return this._indentStart;
    }

    /**
     * Gets the position at which the current indent started.
     */
    public get indentStartPos(): number {
        return this._indentStart ? this._indentStart.startPos : this._linePos;
    }

    /**
     * Gets the current indent depth relative to the current tip.
     */
    public get indent(): number {
        return this._indent;
    }

    /**
     * Gets the current tip.
     */
    public get tip(): Block | undefined {
        return this._tip;
    }

    /**
     * Gets a value indicating whether the current line is blank.
     */
    public get blank(): boolean {
        return this._blank;
    }

    /**
     * Gets the default parsers used to parse block syntax.
     */
    public static getDefaultBlockSyntaxParsers(): ReadonlyArray<IBlockSyntaxParser<Block>> {
        return [
            DocumentParser,
            DocBlockTagParser,
            MarkdownBlockQuoteParser,
            MarkdownHeadingParser,
            MarkdownCodeBlockParser,
            MarkdownHtmlBlockParser,
            MarkdownThematicBreakParser,
            MarkdownListParser,
            MarkdownListItemParser,
            MarkdownParagraphParser
        ];
    }

    public parse(): Document {
        this._line = 0;
        this._linePos = 0;
        this._indentStart = undefined;
        this._indent = 0;
        this._blank = false;
        this._lastNextNonIndentPos = -1;
        this._previousLineEnd = 0;
        this._root = new Document({ pos: 0, text: this.scanner.text });
        this._tip = this._root;
        this._lastMatch = this._root;
        this._hasUnmatchedBlocks = false;

        // prime the scanner with the first token
        this.scanner.scan();
        while (this.scanner.token() !== Token.EndOfFileToken) {
            this._parseLine();
            if (!Token.isLineEnding(this.scanner.token())) {
                throw new Error("Expected to be at end of line or end of input.");
            }
            this._previousLineEnd = this.scanner.startPos;
            this._linePos = this.scanner.pos;
            this._line++;
            this.scanner.scan();
        }
        while (this._tip) {
            this.finish(this._tip, this._previousLineEnd);
        }
        this._processInlines(this._root);
        return this._root;
    }

    private _parseLine(): void {
        let block: Block | undefined = this._root;
        this._lastTip = this._tip;

        // step 1: attempt to continue any open blocks that can accept the current line
        let nextBlock: Block | undefined = block.lastChildBlock;
        while (nextBlock && !this.getParserState(nextBlock).closed) {
            block = nextBlock;
            switch (this._continue(block)) {
                case ContinueResult.Matched:
                    // this.acceptIndent();
                    nextBlock = block.lastChildBlock;
                    continue;
                case ContinueResult.Finished:
                    return;
                case ContinueResult.Unmatched:
                    block = block.parentBlock || this._root;
                    break;
            }
            break;
        }

        this._hasUnmatchedBlocks = block !== this._lastTip;
        this._lastMatch = block;

        // step 2: if the line hasn't already been consumed, start a new block if possible
        let matchedLeaf: boolean = block.kind !== SyntaxKind.MarkdownParagraph && this._acceptsLines(block);
        while (block && !matchedLeaf) {
            switch (this._start(block)) {
                case StartResult.Container:
                    // this.acceptIndent();
                    block = this._tip;
                    continue;
                case StartResult.Leaf:
                    // this.acceptIndent();
                    block = this._tip;
                    matchedLeaf = true;
                    break;
            }
            break;
        }

        // step 3: any remaining content is inline content to be parsed later
        if (this._hasUnmatchedBlocks && !this._blank && this._tip && this._tip.kind === SyntaxKind.MarkdownParagraph) {
            // continue a paragraph
            this._acceptLine(this._tip);
        } else if (block) {
            this.finishUnmatchedBlocks();

            if (this._blank) {
                const lastChild: Block | undefined = block.lastChildBlock;
                if (lastChild) {
                    this.getParserState(lastChild).lastLineIsBlank = true;
                }
            }

            const lastLineIsBlank: boolean = this._blank &&
                !(block.kind === SyntaxKind.MarkdownBlockQuote ||
                    block.kind === SyntaxKind.MarkdownCodeBlock && !!(block as MarkdownCodeBlock).codeFence ||
                    block.kind === SyntaxKind.MarkdownListItem && !block.firstChild && this.lineMap.lineAt(block.pos) === this._line);

            let ancestor: Block | undefined = block;
            while (ancestor) {
                this.getParserState(ancestor).lastLineIsBlank = lastLineIsBlank;
                ancestor = ancestor.parentBlock;
            }

            this._acceptLine(block);
        } 

        // if for some reason we are not at the end of the line, skip to the end of the line
        this.scanner.scanLine();
    }

    private _acceptLine(block: Block): void {
        const blockSyntaxParser: IBlockSyntaxParser<Block> | undefined = this._blockSyntaxParserMap.get(block.kind);
        if (blockSyntaxParser && blockSyntaxParser.acceptLine) {
            blockSyntaxParser.acceptLine(this, block);
        } else if (!Token.isLineEnding(this.scanner.token()) && !this._blank) {
            const pos: number = this.indentStartPos;
            const block: MarkdownParagraph = this.pushBlock(new MarkdownParagraph({ pos, end: pos }));
            MarkdownParagraphParser.acceptLine(this, block);
        }
        this.scanner.scanLine();
    }

    private _start(container: Block): StartResult {
        this.nextNonIndent();
        for (const blockSyntaxParser of this._blockSyntaxParsers) {
            const startResult: StartResult = blockSyntaxParser.tryStart(this, container);
            if (startResult !== StartResult.Unmatched) {
                return startResult;
            }
        }
        return StartResult.Unmatched;
    }

    private _continue(block: Block): ContinueResult {
        this.nextNonIndent();
        const blockSyntaxParser: IBlockSyntaxParser<Block> | undefined = this._blockSyntaxParserMap.get(block.kind);
        return blockSyntaxParser ? blockSyntaxParser.tryContinue(this, block) : ContinueResult.Unmatched;
    }

    public finish(block: Block, end: number = this._previousLineEnd): void {
        const parent: Block | undefined = block.parentBlock;
        this.getParserState(block).closed = true;
        block.end = end;
        try {
            const blockSyntaxParser: IBlockSyntaxParser<Block> | undefined = this._blockSyntaxParserMap.get(block.kind);
            if (blockSyntaxParser) {
                blockSyntaxParser.finish(this, block);
            }
        } finally {
            this._tip = parent;
        }
    }

    public finishUnmatchedBlocks(): void {
        if (this._hasUnmatchedBlocks) {
            while (this._lastTip && this._lastTip !== this._lastMatch) {
                const parent: Block | undefined = this._lastTip.parentBlock;
                this.finish(this._lastTip);
                this._lastTip = parent;
            }
            this._hasUnmatchedBlocks = false;
        }
    }

    public pushBlock<T extends Block>(block: T, pos?: number, end?: number): T {
        if (pos !== undefined) {
            block.pos = pos;
            if (end !== undefined) {
                block.end = end;
            }
        }
        while (this._tip && !this._tip.canHaveChild(block)) {
            const parent: Block | undefined = this._tip.parentBlock;
            this.finish(this._tip);
            this._tip = parent;
        }
        if (this._tip) {
            this._tip.appendChild(block);
            this._tip = block;
        }
        return block;
    }

    public setTip<T extends Block>(block: T): T {
        this._tip = block;
        return block;
    }

    public nextNonIndent(): void {
        // check whether we've advanced the scanner.
        if (this._lastNextNonIndentPos === this.scanner.startPos) {
            return;
        }
        const column: number = this.scanner.startColumn;
        this._indentStart = this.scanner.getState();
        while (Token.isIndentCharacter(this.scanner.token())) {
            this.scanner.scan();
        }
        this._blank = Token.isLineEnding(this.scanner.token());
        this._indent = this.scanner.startColumn - column;
        this._lastNextNonIndentPos = this.scanner.startPos;
    }

    public acceptIndent(): void {
        this._blank = false;
        this._indentStart = undefined;
        this._indent = 0;
        this._lastNextNonIndentPos = -1;
    }

    public retreatToIndentStart(): void {
        this.scanner.setState(this._indentStart);
        this.acceptIndent();
    }

    public parseReferences(block: MarkdownParagraph): void {
        const content: ContentWriter | undefined = this.getParserState(block).content;
        if (content) {
            const parser: InlineParser = new InlineParser(this._root, content.toString(), content.mappings);
            for (const linkReference of parser.parseReferences()) {
                block.insertSiblingBefore(linkReference);
            }

            const scanner: Scanner = parser.scanner;
            if (scanner.token() === Token.EndOfFileToken) {
                this.getParserState(block).content = undefined;
                return;
            }

            if (scanner.startPos === 0) {
                return;
            }
            
            let sourcePos: number = scanner.startPos;
            const newContent: ContentWriter = new ContentWriter();
            newContent.addMapping(scanner.startPos);
            for (const sourceSegment of content.mappings) {
                if (sourceSegment.sourcePos > sourcePos) {
                    newContent.write(scanner.slice(sourcePos, sourceSegment.sourcePos));
                    newContent.addMapping(sourceSegment.sourcePos);
                    sourcePos = sourceSegment.sourcePos;
                }
            }
            newContent.write(scanner.slice(sourcePos));
            this.getParserState(block).content = newContent;
        }
    }

    private _acceptsLines(block: Node): boolean {
        const parser: IBlockSyntaxParser<Block> | undefined = this._blockSyntaxParserMap.get(block.kind);
        return parser && !!parser.acceptLine || false;
    }

    public endsWithBlankLine(block: Block): boolean {
        let current: Block | undefined = block;
        while (current) {
            if (this.getParserState(current).lastLineIsBlank) {
                return true;
            }
            if (!this.getParserState(current).lastLineChecked) {
                this.getParserState(current).lastLineChecked = true;
                if (current.isList()) {
                    current = current.lastChildListItem;
                    continue;
                }
                if (current.isListItem()) {
                    current = current.lastChildBlock;
                    continue;
                }
            }
            break;
        }
        return false;
    }

    private _processInlines(block: Block): void {
        if (block.isInlineContainer()) {
            const content: ContentWriter | undefined = this.getParserState(block).content;
            if (content && content.length) {
                const parser = new InlineParser(this._root, content.toString(), content.mappings);
                parser.parse(block);
            }
            this.getParserState(block).content = undefined;
        } else {
            block.forEachChild(child => {
                if (child.isBlock()) {
                    this._processInlines(child);
                }
            });
        }
    }
}
