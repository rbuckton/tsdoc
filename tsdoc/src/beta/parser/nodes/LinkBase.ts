import { Inline, IInlineParameters } from "./Inline";
import { MarkdownLinkDestination } from "./MarkdownLinkDestination";
import { MarkdownLinkTitle } from "./MarkdownLinkTitle";
import { MarkdownLinkLabel } from "./MarkdownLinkLabel";
import { Syntax } from "./Syntax";
import { Node } from "./Node";
import { MarkdownUtils } from "../utils/MarkdownUtils";
import { MarkdownLinkReference } from "./MarkdownLinkReference";
import { IParserState } from "../ParserBase";

export interface ILinkBaseParameters extends IInlineParameters {
    destination?: MarkdownLinkDestination;
    title?: MarkdownLinkTitle;
    label?: MarkdownLinkLabel;
}

export abstract class LinkBase extends Inline {
    private _destinationSyntax: MarkdownLinkDestination | undefined;
    private _titleSyntax: MarkdownLinkTitle | undefined;
    private _labelSyntax: MarkdownLinkLabel | undefined;
    private _resolvedReference: MarkdownLinkReference | undefined;
    private _documentVersion: number = -1;

    public constructor(parameters?: ILinkBaseParameters) {
        super(parameters);
        this.attachSyntax(this._destinationSyntax = parameters && parameters.destination);
        this.attachSyntax(this._titleSyntax = parameters && parameters.title);
        this.attachSyntax(this._labelSyntax = parameters && parameters.label);
    }

    public get destinationSyntax(): MarkdownLinkDestination | undefined {
        return this._destinationSyntax;
    }

    public get titleSyntax(): MarkdownLinkTitle | undefined {
        return this._titleSyntax;
    }

    public get labelSyntax(): MarkdownLinkLabel | undefined {
        return this._labelSyntax;
    }

    public get href(): string {
        if (this.destinationSyntax) return this.destinationSyntax.href;
        const linkReference: MarkdownLinkReference | undefined = this._resolveLinkReference();
        return linkReference ? linkReference.href : "";
    }

    public set href(value: string) {
        this._applyLinkReference();
        this._ensureDestinationSyntax().href = value;
    }

    public get title(): string | undefined {
        if (this.titleSyntax) return this.titleSyntax.text;
        const linkReference: MarkdownLinkReference | undefined = this._resolveLinkReference();
        return linkReference ? linkReference.title : undefined;
    }

    public set title(value: string | undefined) {
        this._applyLinkReference();
        if (value === undefined) {
            const titleSyntax: MarkdownLinkTitle | undefined = this._titleSyntax;
            if (titleSyntax) {
                this._titleSyntax = undefined;
                this.detachSyntax(titleSyntax);
            }
        } else {
            this._ensureTitleSyntax().text = value;
        }
    }

    public get label(): string | undefined {
        return this._labelSyntax ? this._labelSyntax.text : undefined;
    }

    public set label(value: string | undefined) {
        if (value === undefined) {
            const labelSyntax: MarkdownLinkLabel | undefined = this._labelSyntax;
            if (labelSyntax) {
                this._labelSyntax = undefined;
                this.detachSyntax(labelSyntax);
            }
        } else {
            this._ensureLabelSyntax().text = value;
        }
        this._maybeInvalidateResolvedReference();
    }

    /** @override */
    public isInlineContainer(): true {
        return true;
    }

    /** @override */
    public isLink(): true {
        return true;
    }

    /**
     * Iterates through each syntactic and content element of this node.
     * @param cb The callback to execute for each node. If the callback returns a value other than `undefined`, iteration
     * stops and that value is returned.
     * 
     * @override
     */
    public forEachNode<A extends any[], T>(cb: (node: Node, ...args: A) => T | undefined, ...args: A): T | undefined {
        // inline content comes before other syntax
        return this.forEachChild(cb, ...args)
            || this.forEachSyntax(cb, ...args);
    }

    /** @override */
    protected getSyntax(): ReadonlyArray<Syntax | undefined> {
        return [
            this.destinationSyntax,
            this.titleSyntax,
            this.labelSyntax
        ];
    }

    private _ensureDestinationSyntax(): MarkdownLinkDestination {
        if (!this._destinationSyntax) {
            this.attachSyntax(this._destinationSyntax = new MarkdownLinkDestination());
        }
        return this._destinationSyntax;
    }

    private _ensureTitleSyntax(): MarkdownLinkTitle {
        if (!this._titleSyntax) {
            this.attachSyntax(this._titleSyntax = new MarkdownLinkTitle());
        }
        return this._titleSyntax;
    }

    private _ensureLabelSyntax(): MarkdownLinkLabel {
        if (!this._labelSyntax) {
            this.attachSyntax(this._labelSyntax = new MarkdownLinkLabel());
        }
        return this._labelSyntax;
    }

    private _maybeInvalidateResolvedReference(): void {
        if (this._resolvedReference && (!this.ownerDocument || !this.labelSyntax || Node.getVersion(this.ownerDocument) !== this._documentVersion)) {
            this._resolvedReference = undefined;
            this._documentVersion = -1;
        }
    }

    private _resolveLinkReference(): MarkdownLinkReference | undefined {
        this._maybeInvalidateResolvedReference();
        if (!this._resolvedReference && this.ownerDocument) {
            let refLabel: string | undefined;
            if (this.label) {
                refLabel = MarkdownUtils.normalizeLinkReference(this.label);
            } else {
                const state: IParserState | undefined = this.getParserState();
                refLabel = state && state.refLabel;
            }
            if (refLabel) {
                const linkReference: MarkdownLinkReference | undefined = this.ownerDocument.referenceMap.get(refLabel);
                if (linkReference) {
                    this._resolvedReference = linkReference;
                    this._documentVersion = Node.getVersion(this.ownerDocument);
                }
            }
        }
        return this._resolvedReference;
    }

    private _applyLinkReference(): void {
        const linkReference: MarkdownLinkReference | undefined = this._resolveLinkReference();
        if (linkReference) {
            this._ensureDestinationSyntax().href = linkReference.href;
            if (linkReference.title !== undefined) {
                this._ensureTitleSyntax().text = linkReference.title;
            }
            this._labelSyntax = undefined;
            this._maybeInvalidateResolvedReference();
        }
    }
}
