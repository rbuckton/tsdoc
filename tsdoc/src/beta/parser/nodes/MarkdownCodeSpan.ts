import { Inline } from "./Inline";
import { SyntaxKind } from "./SyntaxKind";
import { INodeParameters } from "./Node";
import { TSDocPrinter } from "../TSDocPrinter";
import { StringUtils } from "../utils/StringUtils";

export interface IMarkdownCodeSpanParameters extends INodeParameters {
    backtickCount?: number;
    text?: string;
}

export class MarkdownCodeSpan extends Inline {
    private _backtickCount: number | undefined;
    private _text: string | undefined;

    public constructor(parameters?: IMarkdownCodeSpanParameters) {
        super(parameters);
        this._backtickCount = parameters && parameters.backtickCount;
        this._text = parameters && parameters.text;
    }

    /** @override */
    public get kind(): SyntaxKind.MarkdownCodeSpan {
        return SyntaxKind.MarkdownCodeSpan;
    }

    public get backtickCount(): number {
        return this._backtickCount || 1;
    }

    public set backtickCount(value: number) {
        if (value < 1) throw new RangeError("Argument out of range: value");
        if (this.backtickCount !== value) {
            this.beforeChange();
            this._backtickCount = value;
            this.afterChange();
        }
    }

    public get text(): string {
        return this._text || "";
    }

    public set text(value: string) {
        if (this.text !== value) {
            this.beforeChange();
            this._text = value;
            this.afterChange();
        }
    }

    /** @override */
    protected print(printer: TSDocPrinter): void {
        printer.write(StringUtils.repeat('`', this.backtickCount));
        if (this.text.indexOf('`') >= 0) {
            printer.write(' ');
            printer.write(this.text);
            printer.write(' ');
        } else {
            printer.write(this.text);
        }
        printer.write(StringUtils.repeat('`', this.backtickCount));
    }
}