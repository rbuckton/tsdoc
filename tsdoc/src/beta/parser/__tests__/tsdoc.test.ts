// This file was autogenerated by ~/scripts/generateTests.js and should not be modified.

import { BlockParser } from "../BlockParser";
import { Document } from "../../nodes/Document";
import { SnapshotSerializer } from "./serializers";
import { HtmlEmitter } from "../../emitters/HtmlEmitter";
import { TSDocConfiguration } from "../../../configuration/TSDocConfiguration";

expect.addSnapshotSerializer(SnapshotSerializer.serializer);

function parse(text: string): Document {
    const parser: BlockParser = new BlockParser(new TSDocConfiguration(), text, undefined, true);
    return parser.parse();
}

function emit(document: Document): string {
    const emitter: HtmlEmitter = new HtmlEmitter(
        new TSDocConfiguration(),
        tagName => !/^(title|textarea|style|xmp|iframe|noembed|noframes|script|plaintext)$/i.test(tagName)
    );
    emitter.emit(document);
    return emitter.toString();
}

describe("Inline Tags (extension)", () => {
    it("Basic Inline", () => {
        const markdown: string = "{@foo bar}";
        const expected: string = "<p><span data-tagname=\"@foo\"><em>@foo</em> &mdash; <a href=\"bar\">bar</a></span></p>\n";
        const document: Document = parse(markdown);
        const actual: string = emit(document);
        expect(actual).toBe(expected);
        expect(document).toMatchSnapshot();
    });
    it("Basic Link", () => {
        const markdown: string = "{@link foo}";
        const expected: string = "<p><a href=\"foo\">foo</a></p>\n";
        const document: Document = parse(markdown);
        const actual: string = emit(document);
        expect(actual).toBe(expected);
        expect(document).toMatchSnapshot();
    });
    it("Link with title", () => {
        const markdown: string = "{@link foo|bar}";
        const expected: string = "<p><a href=\"foo\">bar</a></p>\n";
        const document: Document = parse(markdown);
        const actual: string = emit(document);
        expect(actual).toBe(expected);
        expect(document).toMatchSnapshot();
    });
    it("Link with title in table", () => {
        const markdown: string = "|a|\n|-|\n{@link foo\\|bar}|";
        const expected: string = "<table>\n<thead>\n<tr>\n<th>a</th>\n</tr>\n</thead>\n<tbody>\n<tr>\n<td><a href=\"foo\">bar</a></td>\n</tr>\n</tbody>\n</table>\n";
        const document: Document = parse(markdown);
        const actual: string = emit(document);
        expect(actual).toBe(expected);
        expect(document).toMatchSnapshot();
    });
});
describe("Block Tags (extension)", () => {
    it("Basic Block", () => {
        const markdown: string = "@remarks this is a test";
        const expected: string = "<div data-tagname=\"@remarks\"><em>@remarks</em> &mdash;\n<p>this is a test</p>\n</div>\n";
        const document: Document = parse(markdown);
        const actual: string = emit(document);
        expect(actual).toBe(expected);
        expect(document).toMatchSnapshot();
    });
});
describe("Modifier Tags (extension)", () => {
    it("Single Modifier", () => {
        const markdown: string = "@public";
        const expected: string = "<div data-tagname=\"@public\"><em>@public</em></div>\n";
        const document: Document = parse(markdown);
        const actual: string = emit(document);
        expect(actual).toBe(expected);
        expect(document).toMatchSnapshot();
    });
});
describe("Param Tags (extension)", () => {
    it("Basic Parameter", () => {
        const markdown: string = "@param foo bar";
        const expected: string = "<div data-tagname=\"@param\"><em>@param</em> &mdash; <code>foo</code> &mdash;\n<p>bar</p>\n</div>\n";
        const document: Document = parse(markdown);
        const actual: string = emit(document);
        expect(actual).toBe(expected);
        expect(document).toMatchSnapshot();
    });
});