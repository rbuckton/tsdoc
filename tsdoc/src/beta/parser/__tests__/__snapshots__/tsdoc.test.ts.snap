// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block Tags (extension) Basic Block 1`] = `
Document {
  "kind": Document,
  "children": Array [
    TSDocBlockTag {
      "kind": TSDocBlockTag,
      "children": Array [
        MarkdownParagraph {
          "kind": MarkdownParagraph,
          "children": Array [
            Run {
              "kind": Run,
              "text": "this is a test",
              "pos": 9,
              "end": 23,
            },
          ],
          "pos": 9,
          "end": 23,
        },
      ],
      "tagName": "@remarks",
      "pos": 0,
      "end": 23,
    },
  ],
  "pos": 0,
  "end": 23,
}
`;

exports[`Inline Tags (extension) Basic Inline 1`] = `
Document {
  "kind": Document,
  "children": Array [
    MarkdownParagraph {
      "kind": MarkdownParagraph,
      "children": Array [
        TSDocInlineTag {
          "kind": TSDocInlineTag,
          "children": Array [
            Run {
              "kind": Run,
              "text": "bar",
              "pos": 6,
              "end": 9,
            },
          ],
          "pos": 0,
          "end": 10,
        },
      ],
      "pos": 0,
      "end": 10,
    },
  ],
  "pos": 0,
  "end": 10,
}
`;

exports[`Inline Tags (extension) Basic Link 1`] = `
Document {
  "kind": Document,
  "children": Array [
    MarkdownParagraph {
      "kind": MarkdownParagraph,
      "children": Array [
        TSDocInlineTag {
          "kind": TSDocInlineTag,
          "children": Array [
            Run {
              "kind": Run,
              "text": "foo",
              "pos": 7,
              "end": 10,
            },
          ],
          "pos": 0,
          "end": 11,
        },
      ],
      "pos": 0,
      "end": 11,
    },
  ],
  "pos": 0,
  "end": 11,
}
`;

exports[`Inline Tags (extension) Link with title 1`] = `
Document {
  "kind": Document,
  "children": Array [
    MarkdownParagraph {
      "kind": MarkdownParagraph,
      "children": Array [
        TSDocInlineTag {
          "kind": TSDocInlineTag,
          "children": Array [
            Run {
              "kind": Run,
              "text": "bar",
              "pos": 11,
              "end": 14,
            },
          ],
          "pos": 0,
          "end": 15,
        },
      ],
      "pos": 0,
      "end": 15,
    },
  ],
  "pos": 0,
  "end": 15,
}
`;

exports[`Inline Tags (extension) Link with title in table 1`] = `
Document {
  "kind": Document,
  "children": Array [
    GfmTable {
      "kind": GfmTable,
      "children": Array [
        GfmTableRow {
          "kind": GfmTableRow,
          "children": Array [
            GfmTableCell {
              "kind": GfmTableCell,
              "children": Array [
                Run {
                  "kind": Run,
                  "text": "a",
                  "pos": 1,
                  "end": 2,
                },
              ],
              "pos": 1,
              "end": 2,
            },
          ],
          "pos": 0,
          "end": 3,
        },
        GfmTableRow {
          "kind": GfmTableRow,
          "children": Array [
            GfmTableCell {
              "kind": GfmTableCell,
              "children": Array [
                TSDocInlineTag {
                  "kind": TSDocInlineTag,
                  "children": Array [
                    Run {
                      "kind": Run,
                      "text": "bar",
                      "pos": 20,
                      "end": 23,
                    },
                  ],
                  "pos": 8,
                  "end": 24,
                },
              ],
              "pos": 8,
              "end": 24,
            },
          ],
          "pos": 8,
          "end": 25,
        },
      ],
      "pos": 0,
      "end": 25,
    },
  ],
  "pos": 0,
  "end": 25,
}
`;

exports[`Modifier Tags (extension) Single Modifier 1`] = `
Document {
  "kind": Document,
  "children": Array [
    TSDocModifierTag {
      "kind": TSDocModifierTag,
      "pos": 0,
      "end": 7,
    },
  ],
  "pos": 0,
  "end": 7,
}
`;

exports[`Param Tags (extension) Basic Parameter 1`] = `
Document {
  "kind": Document,
  "children": Array [
    TSDocParamTag {
      "kind": TSDocParamTag,
      "children": Array [
        MarkdownParagraph {
          "kind": MarkdownParagraph,
          "children": Array [
            Run {
              "kind": Run,
              "text": "bar",
              "pos": 11,
              "end": 14,
            },
          ],
          "pos": 11,
          "end": 14,
        },
      ],
      "pos": 7,
      "end": 14,
    },
  ],
  "pos": 0,
  "end": 14,
}
`;
